### PR TITLE
Ap/s3 - reworking everything to pull down data from S3

### PR DESCRIPTION
**Pull request by austinhpatton**
We discussed these changes previously, but basically this PR includes a number of changes so that data is now pulled down from S3. Other changes include modifications so that variable names clearly indicate whether each is to be used for the "mcl_test_dataset" or "complete dataset". Changes to the readme reflect this.

### CODE REVIEW

**Review by elizabethmcd — COMMENTED** [review]
I think the actual changes look good and I like the clarification of the `complete_samplesheet.csv` and `mcl_test_samplesheet.csv` for clarifying your versions of "test" and "full" datasets in this case. 

My confusion is why some of these changes are needed when pulling from S3. For example I can either provide `--input "s3://bucket/location/*_{1,2}.fq.gz` for metagenomic samples or `--input local/dir/location/*_{1,2}.fq.gz` or if I have the samplesheet module then it can take in a samplesheet. However any of these options don't change downstream processes. Do you have to make changes because then this changes the orthofinder directory structure depending on where you are inputting data from? Or because essentially you need multiple samplesheets (for input data, and then mcl parameters)?

**Comment by austinhpatton** [comment]
Okay, great! Glad that clarification in the naming convention helps. 

As for why those changes are needed... I don't have a very concrete answer unfortunately. But, my understanding is that it has to do with the use of the samplesheet_check.py script, which seems to strip the S3 paths from the filename (if I included a column containing the full s3 filepath to that fasta in the samplesheet) when doing the check. 

Because that its the "checked" samplesheet that anything downstream is using, nextflow no longer recognizes them as filepaths and can't download the data.... That's why I ended up having to tweak things to cajole nextflow back into recognizing that I'm specifying paths to s3 and that I want to download those files. 

That's why I'm almost wondering whether I could get away with just assuming that there are no issues with the samplesheet, scrapping the use of that python script and just jumping right into downloaded the data and moving on. Could take some fanangling to get it right still...

File: modules/local/mcl_test_samplesheet_check.nf [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script: // This python script is bundled with the pipeline, in phylorthology/bin/
+    """
+    check_samplesheet.py \\
+        $mcl_test_samplesheet \\
+        mcl_test_samplesheet.valid.csv
+    
+    # Copy the S3 files to the publishdir, adding this as a column to the
```
[the remark this answers is not in the document]
Comment (mertcelebi) [reply modules/local/mcl_test_samplesheet_check.nf:34]: Actually, looking at `bin/check_samplesheet`: I don't think the issue is there. The issue is probably upstream around where the files are staged and how. So, we should investigate that.

File: modules/local/mcl_test_samplesheet_check.nf [hunk 2]
```diff
@@ -0,0 +1,57 @@
+process MCL_TEST_SAMPLESHEET_CHECK {
```
Comment (mertcelebi) [inline modules/local/mcl_test_samplesheet_check.nf:1]: Why do we need a separate process? Is it different from what the other `samplesheet_check` module does?

**Review by mertcelebi — CHANGES_REQUESTED** [review]
There are couple problems with this change:
* I don't think the separate `s3_dir` param is the right approach and the cascading changes to the `input_check` subworkflow should not be necessary.
* You clearly identified a problem: you're unable to stage files from S3 correctly. But your fix is more of a hack than identifying the root cause. 
* Why does being hacky here a bad idea? Mainly because the file I/O is a pretty essential part of the workflow process and if we only fix symptoms of the problem and not the root cause, we'll very likely run into similar issues and have to come up with similar hacks.
* Given the way Nextflow stages files, you don't have to do anything special to use S3 URIs or local file paths to specify your inputs. You can see in the [DLC](https://github.com/Arcadia-Science/trove-deeplabcut) pipeline, there's no explicit reference to S3 in the code. 
* So, we should identify the root problem and fix that. I am happy to pair program this and try to figure it out as it could be due to many many reasons.

File: test/complete_samplesheet.csv [hunk 3]
```diff
@@ -0,0 +1,7 @@
+species,file,taxonomy,shallow,broad,mode,uniprot,mcl_test
+Entamoeba_histolytica,Entamoeba_histolytica-test-proteome.fasta,Amoebozoa,eukaryota_odb10,eukaryota_odb10,proteins,true,false
```
[the remark this answers is not in the document]
Comment (austinhpatton) [reply test/complete_samplesheet.csv:2]: You can see in the updated version, but I was still unable to deal with the S3 path recognition loss due to reading sample info with fromCSV. The solution I found to keep everything together was to specify a data_dir parameter in the parameters.json file (which could be local or s3) and then combine with the filename listed in the samplesheet. 

Not ideal, but the fromCSV quirk of losing AWS cli recognition demanded (from what I could tell) a somewhat hacky solution. This one is a least a bit more generalizable than before, requiring only the one samplesheet.

File: modules/local/mcl_test_samplesheet_check.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (austinhpatton) [reply modules/local/mcl_test_samplesheet_check.nf:34]: We talked about this before, but this issue is caused by the fromCSV nextflow operator losing recognition of the path as one to S3 and so no longer using the AWS CLI. I've also deleted this mcl-test specific samplesheet check module.

File: modules/local/mcl_test_samplesheet_check.nf (hunk 2 above)
Comment (austinhpatton) [reply modules/local/mcl_test_samplesheet_check.nf]: Nope - deleted it.

File: modules/local/mcl_test_samplesheet_check.nf [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    tag "$mcl_test_samplesheet"
+
+    conda (params.enable_conda ? "conda-forge::python=3.9.5" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
```
[the remark this answers is not in the document]
Comment (austinhpatton) [reply modules/local/mcl_test_samplesheet_check.nf]: I'll  look again, but from what I remember there was no 3.9.5 docker - don't know if there's a 3.9--1 for conda. These are the environments that were provided by default with this module.

File: modules/local/mcl_test_samplesheet_check.nf [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        'https://depot.galaxyproject.org/singularity/python:3.9--1' :
+        'quay.io/biocontainers/python:3.9--1' }"
+    
+    // TODO: THESE PUBLISHDIRS MIGHT NOT BE NECESSESARY?
```
[the remark this answers is not in the document]
Comment (austinhpatton) [reply modules/local/mcl_test_samplesheet_check.nf]: I think we can ignore this now.

File: subworkflows/local/input_check.nf [hunk 6]
```diff
@@ -34,6 +48,6 @@ def create_prots_channel(LinkedHashMap row) {
 
     // add path(s) of the proteome file to the meta map
     def prots_meta = []
-        prots_meta = [ meta, [ file(row.file) ] ] 
+        prots_meta = [ meta, [ file(row.fpath) ] ] 
     return prots_meta
 }
```
[the remark this answers is not in the document]
Comment (austinhpatton) [reply subworkflows/local/input_check.nf]: Fixed.

File: modules/local/samplesheet_check.nf [hunk 7]
```diff
@@ -1,30 +1,65 @@
 process SAMPLESHEET_CHECK {
-    tag "$samplesheet"
+    tag "$complete_samplesheet"
 
     conda (params.enable_conda ? "conda-forge::python=3.9.5" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/python:3.9--1' :
         'quay.io/biocontainers/python:3.9--1' }"
+    
+    // TODO: FOR SOME REASON THIS MODULE OUTPUTS TO "PIPELINE_INFO" BUT THE MCL_SAMPLESHEET_CHECK OUTPUTS TO ITS OWN DIRECTORY.
```
[the remark this answers is not in the document]
Comment (austinhpatton) [reply modules/local/samplesheet_check.nf:9]: Okay - eventually I'll want to migrate some hard-coded parameter choices to this config file (might be cleaner than including them all in the parameters.json file, though maybe doing so would be more intuitive for new nextflow users?)

File: modules/local/mcl_test_samplesheet_check.nf (hunk 4 above)
[the remark this answers is not in the document]
Comment (austinhpatton) [reply modules/local/mcl_test_samplesheet_check.nf]: Okay, nevermind - made them all specify 3.9--1 now.

**Review by mertcelebi — APPROVED** [review]
Austin and I had a conversation about this PR. He addressed the immediate issues in the PR, but I'll tackle the rest of this going forward. So we'll merge this PR and branch off so he can work on other things and I can have a working version myself.

**Pull request merged**